### PR TITLE
Add root-ca configuration to our dataplane agents for https proxies

### PR DESCRIPTION
**Pull request by RazOmessi**

### CODE REVIEW

File: cbcontainers/communication/gateway/api_gateway.go [hunk 1]
```diff
@@ -21,9 +20,34 @@ type ApiGateway struct {
...
+
+		tlsConfig.RootCAs = rootCAs
+	}
+
+	client.SetTLSClientConfig(&tlsConfig)
+	return client, nil
+}
+
+func NewApiGateway(account, cluster string, accessToken string, scheme, host string, port int, adapter string,
+	tlsInsecureSkipVerify bool, rootCAsBundle []byte) (*ApiGateway, error) {
```
Comment (asankov-cb) [inline cbcontainers/communication/gateway/api_gateway.go:44]: this method has quite a lot of parameters already. might be a good idea to extract them into a struct. something like:
```go
func NewApiGateway(cfg *GatewayConfig) (*ApiGateway, error)
```

might not be now, I interact with this in #52 , so I can do it will my changes. WDYT?

Comment (asankov-cb) [reply cbcontainers/communication/gateway/api_gateway.go:44]: nvm this comment now, I just saw the `Builder`. I think it's neat.

File: cbcontainers/state/common/dataplane_consts.go [hunk 2]
```diff
@@ -8,12 +8,17 @@ const (
...
+	DataPlaneConfigmapClusterKey        = "Cluster"
+	DataPlaneConfigmapApiSchemeKey      = "ApiScheme"
+	DataPlaneConfigmapApiHostKey        = "ApiHost"
+	DataPlaneConfigmapApiPortKey        = "ApiPort"
+	DataPlaneConfigmapApiAdapterKey     = "ApiAdapter"
+	DataPlaneConfigmapTlsSkipVerifyKey  = "TLS.SkipVerify"
+	DataPlaneConfigmapTlsRootCAsPathKey = "TLS.RootCAsPath"
+
+	DataPlaneConfigmapTlsRootCAsDirPath = "/etc/gateway-certs"
+	DataPlaneConfigmapTlsRootCAsFilePath = "root.pem"
```
Comment (asankov-cb) [inline cbcontainers/state/common/dataplane_consts.go:21]: this is not gofmt-ed
```suggestion
	DataPlaneConfigmapTlsRootCAsDirPath  = "/etc/gateway-certs"
	DataPlaneConfigmapTlsRootCAsFilePath = "root.pem"
```

File: cbcontainers/state/common/volumes.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
+package common
```
Comment (asankov-cb) [inline cbcontainers/state/common/volumes.go:1]: do you think this logic deserves it's own package, e.g. `common/volumes`?

File: cbcontainers/state/common/volumes.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+func GetVolumeIndexForName(templatePodSpec *coreV1.PodSpec, name string) int {
+	for i, volume := range templatePodSpec.Volumes {
+		if volume.Name == name {
+			return i
+		}
+	}
+
+	volume := coreV1.Volume{Name: name}
+	templatePodSpec.Volumes = append(templatePodSpec.Volumes, volume)
```
Comment (asankov-cb) [inline cbcontainers/state/common/volumes.go:18]: I don't like how the method name is `GetSomething` but in the same time it mutates its arguments. Can we come up with a better name for it? also - it seems that it does 2 things - get the index of a volume OR add a new volume with the same name.

maybe `GetOrCreateVolumeForName` and a good documentation should be enough

File: cbcontainers/state/common/volumes.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+			configMapSource.Items[i].Path = path
+			return
+		}
+	}
+
+	item := coreV1.KeyToPath{Key: key, Path: path}
+	configMapSource.Items = append(configMapSource.Items, item)
+}
+
+func MutateVolumesToIncludeRootCasVolume(templatePodSpec *coreV1.PodSpec) {
```
Comment (asankov-cb) [inline cbcontainers/state/common/volumes.go:34]:
```suggestion
func MutateVolumesToIncludeRootCAsVolume(templatePodSpec *coreV1.PodSpec) {
```

File: cbcontainers/state/common/volumes.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	for i, volumeMount := range container.VolumeMounts {
+		if volumeMount.Name == name {
+			return i
+		}
+	}
+
+	volumeMount := coreV1.VolumeMount{Name: name}
+	container.VolumeMounts = append(container.VolumeMounts, volumeMount)
+	return len(container.VolumeMounts) - 1
+}
```
Comment (asankov-cb) [inline cbcontainers/state/common/volumes.go:57]: same comment as in `GetVolumeIndexForName`

File: cbcontainers/state/common/volumes.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+func MutateVolumeMount(container *coreV1.Container, volumeMountIndex int, mountPath string, readOnly bool) {
+	container.VolumeMounts[volumeMountIndex].MountPath = mountPath
+	container.VolumeMounts[volumeMountIndex].ReadOnly = readOnly
+}
+
+func MutateVolumeMountToIncludeRootCasVolumeMount(container *coreV1.Container) {
```
Comment (asankov-cb) [inline cbcontainers/state/common/volumes.go:64]:
```suggestion
func MutateVolumeMountToIncludeRootCAsVolumeMount(container *coreV1.Container) {
```

File: cbcontainers/state/hardening/objects/enforcer_deployment.go [hunk 8]
```diff
@@ -19,8 +19,9 @@ const (
 	DesiredContainerPortName  = "https"
 	DesiredContainerPortValue = 443
 
-	DesiredTlsSecretVolumeName      = "cert"
-	DesiredTlsSecretVolumeMountPath = "/etc/octarine-certificates"
+	DesiredTlsSecretVolumeName          = "cert"
+	DesiredTlsSecretVolumeMountPath     = "/etc/octarine-certificates"
+	DesiredTlsSecretVolumeMountReadOnly = true
```
Comment (asankov-cb) [inline cbcontainers/state/hardening/objects/enforcer_deployment.go:24]: these should be `DesiredTLS...`, but you don't have to change them now, since 2 of them are old

**Review by asankov-cb — CHANGES_REQUESTED** [review]

File: cbcontainers/state/cluster/objects/monitor_deployment.go [hunk 9]
```diff
@@ -74,11 +74,20 @@ func (obj *MonitorDeploymentK8sObject) MutateClusterChildK8sObject(k8sObject cli
...
 
 	return nil
 }
 
+func (obj *MonitorDeploymentK8sObject) mutateVolumes(templatePodSpec *coreV1.PodSpec) {
+	if templatePodSpec.Volumes == nil || len(templatePodSpec.Volumes) != 1 {
+		templatePodSpec.Volumes = make([]coreV1.Volume, 0)
+	}
+
+	commonState.MutateVolumesToIncludeRootCasVolume(templatePodSpec)
```
Comment (ltsonov-cb) [inline cbcontainers/state/cluster/objects/monitor_deployment.go:88]: [Minor] `RootCas` -> `RootCAs`

File: cbcontainers/state/common/volumes.go (hunk 4 above)
Comment (ltsonov-cb) [reply cbcontainers/state/common/volumes.go:18]: Agreed, I'd prefer if it returns -1 when the volume is not found so the caller can decide what to do instead

I guess it also does not have to be exported ?

File: cbcontainers/communication/gateway/api_gateway.go [hunk 10]
```diff
@@ -21,9 +20,34 @@ type ApiGateway struct {
...
+
+	if len(rootCAsBundle) > 0 {
+		rootCAs := x509.NewCertPool()
+		ok := rootCAs.AppendCertsFromPEM(rootCAsBundle)
+		if !ok {
+			return nil, fmt.Errorf("failed to parse root certificates")
+		}
+
+		tlsConfig.RootCAs = rootCAs
+	}
```
Comment (ltsonov-cb) [inline cbcontainers/communication/gateway/api_gateway.go:37]: Should we output some form of warning when `InsecureSkipVerify` is true since that is a pretty big deal? 

Or at least when `InsecureSkipVerify` is true *and* `RootCAs` is also set, because I don't see how this makes sense

**Comment by ltsonov-cb** [comment]
A few overall comments:

Do we need to make the gateway TLS config mandatory? What if I don't want do deploy any proxies (e.g. in our own clusters locally) - how would that work? 

It feels like this feature should be an opt-in advanced setting for specific scenarios but it is not clear to from the MR alone how do I run the agent with/without it and what changes. 

Also, if it is valid to run without any of the gateway settings (like root CAs, etc.) - then should we mount the volume only when needed?

File: cbcontainers/state/common/volumes.go (hunk 4 above)
Comment (asankov-cb) [reply cbcontainers/state/common/volumes.go:18]: it is used outside of the package. 

I am OK with the function doing 2 things, I just think that we should name it better so that it is clear what exactly does it do.

File: cbcontainers/state/common/volumes.go (hunk 3 above)
Comment (RazOmessi) [reply cbcontainers/state/common/volumes.go:1]: I agree we might want find a better name. currently all the files in that pkg are doing some common mutations, so maybe we need to change the pkg to common/mutations/* wdyt?

File: cbcontainers/state/common/volumes.go (hunk 4 above)
Comment (RazOmessi) [reply cbcontainers/state/common/volumes.go:18]: wdyt about `EnsureAndGetVolumeIndexForName`?

File: cbcontainers/state/common/volumes.go (hunk 3 above)
Comment (asankov-cb) [reply cbcontainers/state/common/volumes.go:1]: I think a better separation would be `common/volumes`, `common/otherk8sobject`, etc. 

e.g. separation by k8s object, not by action.

also if we take a look at the consumers:
```go
import "github.com/vmware/cbcontainers-operator/.../common/volumes"

volumes.GetOrCreateVolumeByName()
```

```go
import "github.com/vmware/cbcontainers-operator/.../common/mutations"

mutations.GetOrCreateVolume()
```

both look OK, but the first one can be shortened to:
```go
import "github.com/vmware/cbcontainers-operator/.../common/volumes"

volumes.GetOrCreate()
```

TLDR: I prefer `common/volumes`, but I don't hold that opinion pretty strongly.

File: cbcontainers/state/common/volumes.go (hunk 4 above)
Comment (asankov-cb) [reply cbcontainers/state/common/volumes.go:18]: I like it. this is a long name and it describes the action pretty clearly.

a shorter one could be `EnsureVolumeIndexForName` and in the godoc we can write that the return value is actually the one we are ensuring in the first place.

I think both are better than the current one, so whatever you fancy more.

**Review by asankov-cb — APPROVED** [review]

File: cbcontainers/state/common/volumes.go (hunk 3 above)
Comment (RazOmessi) [reply cbcontainers/state/common/volumes.go:1]: Hmm I agree we can go for it - so we will have under common
* image
* volumes
* envvars
* probes
* tests
* consts

Comment (asankov-cb) [reply cbcontainers/state/common/volumes.go:1]: sounds ok, except for:
- what is `tests` for?
- not sure if we need a separate package for `consts`. each type-specific const can be in the type-specific package. the common constants can be in the root.

also - don't feel pressured to do it now. we can just open a Jira about it.

Comment (RazOmessi) [reply cbcontainers/state/common/volumes.go:1]: cool, i'll open a ticket :)
